doc/error_handling_guidelines.md

```diff
@@ -0,0 +1,130 @@
+```
+__init__(self, error_msg, recommendation=None):
+```
+- `error_msg`: _string_, _required_. A clear message shown to users what the error is.
+- `recommendation`: _string_ or _list_, _optional_. Recommendations telling users what action to take.
+
+### Add a New Error Type
+
+If there is not a proper error type for your case and the error is general enough, consider defining a new error type in [azure/cli/core/azclierror.py](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli-core/azure/cli/core/azclierror.py). The defined error type should be inherited from one of the errors define in the second layer (`UserFault`, `ClientError`, `ServiceError`).
+
+For example, a new error type can be defined in this way.
+```
+class NewErrorTypeName(UserFault):
+    """ A description of new error type. """
+    pass
+```
+
+
+## Error Message
+
+A clear and actionable error message is very important when raising an error, so make sure your error message describe clearly what the error is and provide users what they need to do if possible.
+
+A general pattern is provide here, keep it in mind when you write an error message.
+
+1. __What the error is.__
+2. __Why it happens.__
+3. __What users need to do to fix it.__
+
+Below are the specific DOs and DON'Ts when writing the error messages. PRs violates rules here will be rejected.
+
+__DOs__
+- Use the capital letter ahead of an error message.
+
+__DON'Ts__
+- Do not control the style of an error message. (e.g, the unnecessary `'\n'` and the colorization.)
+- Do not include the error type info in an error message. (e.g, `usage error: --ids | --name [--resource-group]`)
+- Do not use a formula-like or a programming expression in the error message. (e.g, `Parameter 'resource_group_name' must conform to the following pattern: '^[-\\w\\._\\(\\)]+$'`)
+- Do not use ambiguous expressions which mean nothing to users. (e.g, `Something unexpected happens.`)
+
+
+## Error Recommendation
+
+When necessary, it is highly suggested for command group authors to provide recommendations for users to resolve the errors they are encountering. It is also suggested that we split the error message itself from the recommendations, which can be done either by specifying a recommendation property when initiating an error type or using the `set_recommendation` function after an error is initiated. The recommendations you provide will be printed right below the error message, one recommendation in a new line.
+
+
+```
+from azure.cli.core.azclierror import MutuallyExclusiveArgumentError
+
+error_msg = 'Please specify all of (--publisher, --offer, --sku, --version), or --urn'
+recommendation = 'Try to use --urn publisher:offer:sku:version only'
+raise MutuallyExclusiveArgumentError(err_msg, recommendation)
```

Comment:
How to add multiple recommendations?

Reply:
good catch!
I will add a description for this but may not a code snippet because I can not find the case we need to provide multiple recommendations right now.